src/Interpreters/Cache/FileSegment.cpp

```diff
@@ -781,21 +795,42 @@ bool FileSegment::assertCorrectness() const
     return assertCorrectnessUnlocked(lockFileSegment());
 }
 
-bool FileSegment::assertCorrectnessUnlocked(const FileSegmentGuard::Lock &) const
+bool FileSegment::assertCorrectnessUnlocked(const FileSegmentGuard::Lock & lock) const
 {
-    auto check_iterator = [this](const Priority::IteratorPtr & it)
+    auto throw_logical = [&](const std::string & error)
+    {
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "{}. File segment info: {}", error, getInfoForLogUnlocked(lock));
+    };
+
+    auto check_iterator = [&](const Priority::IteratorPtr & it)
     {
         UNUSED(this);
         if (!it)
             return;
 
         const auto & entry = it->getEntry();
-        UNUSED(entry);
-        chassert(entry->size == reserved_size);
+        if (download_state != State::DOWNLOADING && entry->size != reserved_size)
```

Comment:
This additional check for downloaded state fixes https://github.com/ClickHouse/ClickHouse/issues/55867.

Reply:
just for my information, why it cannot be e.g. `PARTIALLY_DOWNLOADED_NO_CONTINUATION`?

Reply:
In `tryReserve` we first update `entry->size` and right after `reserved_size`, which is done under cache lock (which is priority queue lock). But here in `assertCacheCorrectness` this lock is not taken, so it can see `entry->size != reserved_size`. As `tryReserve` is executing while file segment has `DOWNLOADING` state, therefore I added this check. `PARTIALLY_DOWNLOADED_NO_CONTINUATION` does not apply to the described case.